dulwich/client.py

```diff
@@ -1948,6 +1948,15 @@ def _discover_references(self, service, base_url):
                     % (url, resp.redirect_location, tail)
                 )
             base_url = resp.redirect_location[: -len(tail)]
+            base_url_parsed = urlparse(base_url)
```

Comment:
could this just use urljoin?

```suggestion
            base_url = urllib.parse.urljoin(url, base_url)
```

Reply:
This works for the edge case fixed in that PR but not for some classical  redirection URLs, see failing test below:
```
_________________________________________________________________________________ HttpGitClientTests.test_url_redirect_location __________________________________________________________________________________

self = <dulwich.tests.test_client.HttpGitClientTests testMethod=test_url_redirect_location>

    def test_url_redirect_location(self):
        from urllib3.response import HTTPResponse
    
        test_data = {
            "https://gitlab.com/inkscape/inkscape/": {
                "location": "https://gitlab.com/inkscape/inkscape.git/",
                "redirect_url": "https://gitlab.com/inkscape/inkscape.git/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"fb2bebf4919a011f0fd7cec085443d0031228e76 "
                    b"HEAD\n0000"
                ),
            },
            "https://github.com/jelmer/dulwich/": {
                "location": "https://github.com/jelmer/dulwich/",
                "redirect_url": "https://github.com/jelmer/dulwich/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"3ff25e09724aa4d86ea5bca7d5dd0399a3c8bfcf "
                    b"HEAD\n0000"
                ),
            },
            # check for absolute-path URI reference as location
            "https://codeberg.org/ashwinvis/radicale-sh.git/": {
                "location": "/ashwinvis/radicale-auth-sh/",
                "redirect_url": "https://codeberg.org/ashwinvis/radicale-auth-sh/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"470f8603768b608fc988675de2fae8f963c21158 "
                    b"HEAD\n0000"
                ),
            },
        }
    
        tail = "info/refs?service=git-upload-pack"
    
        # we need to mock urllib3.PoolManager as this test will fail
        # otherwise without an active internet connection
        class PoolManagerMock:
            def __init__(self):
                self.headers = {}
    
            def request(self, method, url, fields=None, headers=None, redirect=True, preload_content=True):
                base_url = url[: -len(tail)]
                redirect_base_url = test_data[base_url]["location"]
                redirect_url = redirect_base_url + tail
                headers = {
                    "Content-Type": "application/x-git-upload-pack-advertisement"
                }
                body = test_data[base_url]["refs_data"]
                # urllib3 handles automatic redirection by default
                status = 200
                request_url = redirect_url
                # simulate urllib3 behavior when redirect parameter is False
                if redirect is False:
                    request_url = url
                    if redirect_base_url != base_url:
                        body = b""
                        headers["location"] = test_data[base_url]["location"]
                        status = 301
    
                return HTTPResponse(
                    body=BytesIO(body),
                    headers=headers,
                    request_method=method,
                    request_url=request_url,
                    preload_content=preload_content,
                    status=status,
                )
    
        pool_manager = PoolManagerMock()
    
        for base_url in test_data.keys():
            # instantiate HttpGitClient with mocked pool manager
            c = HttpGitClient(base_url, pool_manager=pool_manager, config=None)
            # call method that detects url redirection
            _, _, processed_url = c._discover_references(b"git-upload-pack", base_url)
    
            # send the same request as the method above without redirection
            resp = c.pool_manager.request("GET", base_url + tail, redirect=False)
    
            # check expected behavior of urllib3
            redirect_location = resp.get_redirect_location()
    
            if resp.status == 200:
                self.assertFalse(redirect_location)
    
            if redirect_location:
                # check that url redirection has been correctly detected
>               self.assertEqual(processed_url, test_data[base_url]["redirect_url"])
E               AssertionError: 'https://gitlab.com/inkscape/inkscape/' != 'https://gitlab.com/inkscape/inkscape.git/'
E               - https://gitlab.com/inkscape/inkscape/
E               + https://gitlab.com/inkscape/inkscape.git/
E               ?                                     ++++

dulwich/tests/test_client.py:1129: AssertionError
```

Reply:
In fact it does not work either for absolute path URI, previous test was failing at first error, I commented out some tests data to also test with it.
```
_________________________________________________________________________________ HttpGitClientTests.test_url_redirect_location __________________________________________________________________________________

self = <dulwich.tests.test_client.HttpGitClientTests testMethod=test_url_redirect_location>

    def test_url_redirect_location(self):
        from urllib3.response import HTTPResponse
    
        test_data = {
            # "https://gitlab.com/inkscape/inkscape/": {
            #     "location": "https://gitlab.com/inkscape/inkscape.git/",
            #     "redirect_url": "https://gitlab.com/inkscape/inkscape.git/",
            #     "refs_data": (
            #         b"001e# service=git-upload-pack\n00000032"
            #         b"fb2bebf4919a011f0fd7cec085443d0031228e76 "
            #         b"HEAD\n0000"
            #     ),
            # },
            "https://github.com/jelmer/dulwich/": {
                "location": "https://github.com/jelmer/dulwich/",
                "redirect_url": "https://github.com/jelmer/dulwich/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"3ff25e09724aa4d86ea5bca7d5dd0399a3c8bfcf "
                    b"HEAD\n0000"
                ),
            },
            # check for absolute-path URI reference as location
            "https://codeberg.org/ashwinvis/radicale-sh.git/": {
                "location": "/ashwinvis/radicale-auth-sh/",
                "redirect_url": "https://codeberg.org/ashwinvis/radicale-auth-sh/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"470f8603768b608fc988675de2fae8f963c21158 "
                    b"HEAD\n0000"
                ),
            },
        }
    
        tail = "info/refs?service=git-upload-pack"
    
        # we need to mock urllib3.PoolManager as this test will fail
        # otherwise without an active internet connection
        class PoolManagerMock:
            def __init__(self):
                self.headers = {}
    
            def request(self, method, url, fields=None, headers=None, redirect=True, preload_content=True):
                base_url = url[: -len(tail)]
                redirect_base_url = test_data[base_url]["location"]
                redirect_url = redirect_base_url + tail
                headers = {
                    "Content-Type": "application/x-git-upload-pack-advertisement"
                }
                body = test_data[base_url]["refs_data"]
                # urllib3 handles automatic redirection by default
                status = 200
                request_url = redirect_url
                # simulate urllib3 behavior when redirect parameter is False
                if redirect is False:
                    request_url = url
                    if redirect_base_url != base_url:
                        body = b""
                        headers["location"] = test_data[base_url]["location"]
                        status = 301
    
                return HTTPResponse(
                    body=BytesIO(body),
                    headers=headers,
                    request_method=method,
                    request_url=request_url,
                    preload_content=preload_content,
                    status=status,
                )
    
        pool_manager = PoolManagerMock()
    
        for base_url in test_data.keys():
            # instantiate HttpGitClient with mocked pool manager
            c = HttpGitClient(base_url, pool_manager=pool_manager, config=None)
            # call method that detects url redirection
            _, _, processed_url = c._discover_references(b"git-upload-pack", base_url)
    
            # send the same request as the method above without redirection
            resp = c.pool_manager.request("GET", base_url + tail, redirect=False)
    
            # check expected behavior of urllib3
            redirect_location = resp.get_redirect_location()
    
            if resp.status == 200:
                self.assertFalse(redirect_location)
    
            if redirect_location:
                # check that url redirection has been correctly detected
>               self.assertEqual(processed_url, test_data[base_url]["redirect_url"])
E               AssertionError: 'https://codeberg.org/ashwinvis/radicale-sh.git/' != 'https://codeberg.org/ashwinvis/radicale-auth-sh/'
E               - https://codeberg.org/ashwinvis/radicale-sh.git/
E               ?                                           ----
E               + https://codeberg.org/ashwinvis/radicale-auth-sh/
E               ?
```

Reply:
You'd still want to do the manipulation on line 1950, i.e.:

```
diff --git a/dulwich/client.py b/dulwich/client.py
index 128da1beb..dbd46653a 100644
--- a/dulwich/client.py
+++ b/dulwich/client.py
@@ -1947,7 +1947,7 @@ class AbstractHttpGitClient(GitClient):
                     "Redirected from URL %s to URL %s without %s"
                     % (url, resp.redirect_location, tail)
                 )
-            base_url = resp.redirect_location[: -len(tail)]
+            base_url = urljoin(url, resp.redirect_location[: -len(tail)])

         try:
             self.dumb = (
```

Reply:
Indeed, bad copy paste from my side. Much simpler indeed, will update.